File: typescript/src/levels.ts
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: Apache-2.0

/**
 * The canonical severity ladder and the single place a level string becomes a
 * level.
 *
 * Before this module TypeScript had no converter at all: the raw
 * PROVIDE_LOG_LEVEL value was lowercased and handed straight to pino, whose
 * vocabulary is trace|debug|info|warn|error|fatal. WARNING and CRITICAL —
 * both listed as applicable to TypeScript in spec/telemetry-api.yaml — threw
 * "default level:warning must be included in custom levels" at logger
 * construction. Per-module levels had the same fault.
 *
 * See the log_levels section of spec/behavioral_fixtures.yaml for the
 * cross-language contract.
 */

/**
 * Severity, ordered. The numeric value is the rank, so severities compare
 * directly with `<` and `>`.
 *
 * WARNING and FATAL are deliberately absent: they are spellings resolved by
 * {@link tryParseLevel}, not members. Admitting an alias as a member is how
 * Warn and Warning both ended up on the public C# Logger surface.
 */
export enum LogSeverity {
  Trace = 0,
  Debug = 1,
  Info = 2,
  Warn = 3,
  Error = 4,
  Critical = 5,
}

/** Canonical spellings first, then the two aliases. */
const TABLE: Readonly<Record<string, LogSeverity>> = {
  TRACE: LogSeverity.Trace,
  DEBUG: LogSeverity.Debug,
  INFO: LogSeverity.Info,
  WARN: LogSeverity.Warn,
  ERROR: LogSeverity.Error,
  CRITICAL: LogSeverity.Critical,
  WARNING: LogSeverity.Warn,
  FATAL: LogSeverity.Critical,
};

/** Canonical uppercase spelling, indexed by rank. */
const CANONICAL_NAMES: readonly string[] = ['TRACE', 'DEBUG', 'INFO', 'WARN', 'ERROR', 'CRITICAL'];

/** pino's own vocabulary, indexed by rank. CRITICAL lands on pino's fatal. */
const PINO_NAMES: readonly string[] = ['trace', 'debug', 'info', 'warn', 'error', 'fatal'];

/** The canonical uppercase spelling, as it appears on the record. */
export function severityName(severity: LogSeverity): string {
  return CANONICAL_NAMES[severity] as string;
}

/** The pino level name for a severity. */
export function pinoLevelName(severity: LogSeverity): string {
  return PINO_NAMES[severity] as string;
}

/**
 * Resolve a level string, or null when it is not recognised.
 * Trims surrounding whitespace; comparison is case-insensitive.
 */
export function tryParseLevel(text: string | null | undefined): LogSeverity | null {
  if (text === null || text === undefined) return null;
  const found = TABLE[text.trim().toUpperCase()];
  return found === undefined ? null : found;
}

/**
 * Resolve a level string, substituting `fallback` when it is not recognised.
 * The fallback is a parameter rather than a hidden constant so the
 * substitution is visible at the call site.
 */
export function parseLevel(
  text: string | null | undefined,
  fallback: LogSeverity = LogSeverity.Info,
): LogSeverity {
  const parsed = tryParseLevel(text);
  return parsed === null ? fallback : parsed;
}

/** Numeric rank, for threshold comparisons. */
export function levelOrder(text: string | null | undefined): number {
  return parseLevel(text);
}

/**
 * Normalise any accepted spelling to the pino level name it configures.
 * This is the boundary that stops a canonical spelling reaching pino.
 */
export function toPinoLevel(text: string | null | undefined): string {
  return pinoLevelName(parseLevel(text));
}

/** pino's numeric ladder mapped onto the canonical one. */
const PINO_TO_SEVERITY: Readonly<Record<number, LogSeverity>> = {
  10: LogSeverity.Trace,
  20: LogSeverity.Debug,
  30: LogSeverity.Info,
  40: LogSeverity.Warn,
  50: LogSeverity.Error,
  60: LogSeverity.Critical,
};

/**
 * Resolve a pino numeric level to a severity.
 *
 * pino's numbers are the internal representation; the canonical name is what
 * reaches the record. A number outside pino's ladder resolves to INFO, the
 * same fallback the string parser uses.
 */
export function severityFromPino(level: number): LogSeverity {
  return PINO_TO_SEVERITY[level] ?? LogSeverity.Info;
}

File: csharp/src/Provide.Telemetry/Levels.cs
// SPDX-FileCopyrightText: Copyright (C) 2026 provide.io llc
// SPDX-License-Identifier: Apache-2.0
// SPDX-Comment: Part of provide-telemetry.

namespace Provide.Telemetry;

/// <summary>The canonical severity ladder. Order is the enum value.</summary>
/// <remarks>
/// Named LogSeverity rather than the more guessable LogLevel because
/// Microsoft.Extensions.Logging.LogLevel is near-universal in .NET: a
/// Provide.Telemetry.LogLevel would raise CS0104 on every unqualified use in
/// any file importing both namespaces — including this repo's own
/// OpenTelemetryBackend.cs. A less discoverable name costs one doc line; a
/// name collision costs every downstream consumer a compile error.
///
/// WARNING and FATAL are deliberately absent. They are spellings resolved by
/// <see cref="Levels.TryParse"/>, not members. Admitting an alias as a member
/// is how Warn and Warning both ended up on the public Logger surface.
/// </remarks>
public enum LogSeverity
{
    Trace = 0,
    Debug = 1,
    Info = 2,
    Warn = 3,
    Error = 4,
    Critical = 5,
}

/// <summary>The one place a level string becomes a level.</summary>
/// <remarks>
/// Every level comparison in this library resolves through here. Before it,
/// the port carried three private tables that disagreed: Logger.Rank ranked
/// CRITICAL above ERROR, Governance ranked it the same way, and the OTel
/// backend's MapLevel folded unknown levels onto Information while knowing
/// nothing of FATAL. See the log_levels section of
/// spec/behavioral_fixtures.yaml for the cross-language contract.
/// </remarks>
public static class Levels
{
    // Canonical spellings first, then the two aliases. Case-insensitive, so
    // "warn"/"WARN"/"Warn" are one entry rather than three.
    private static readonly Dictionary<string, LogSeverity> Table = new(StringComparer.OrdinalIgnoreCase)
    {
        ["TRACE"] = LogSeverity.Trace,
        ["DEBUG"] = LogSeverity.Debug,
        ["INFO"] = LogSeverity.Info,
        ["WARN"] = LogSeverity.Warn,
        ["ERROR"] = LogSeverity.Error,
        ["CRITICAL"] = LogSeverity.Critical,
        ["WARNING"] = LogSeverity.Warn,
        ["FATAL"] = LogSeverity.Critical,
    };

    private static readonly string[] CanonicalNames =
    [
        "TRACE",
        "DEBUG",
        "INFO",
        "WARN",
        "ERROR",
        "CRITICAL",
    ];

    /// <summary>The canonical uppercase spelling, as it appears on the record.</summary>
    public static string Name(LogSeverity severity) => CanonicalNames[(int)severity];

    /// <summary>
    /// Resolve a level string, reporting whether it was recognised.
    /// </summary>
    /// <remarks>
    /// On failure <paramref name="severity"/> is set to Info rather than left
    /// at default(LogSeverity) — which is Trace, and would silently promote an
    /// unrecognised level to the most verbose setting in the ladder.
    /// </remarks>
    public static bool TryParse(string? text, out LogSeverity severity)
    {
        if (text is not null && Table.TryGetValue(text.Trim(), out var found))
        {
            severity = found;
            return true;
        }
        severity = LogSeverity.Info;
        return false;
    }

    /// <summary>
    /// Resolve a level string, substituting <paramref name="fallback"/> when it
    /// is not recognised. The fallback is a parameter, not a hidden constant,
    /// so the substitution is visible at the call site.
    /// </summary>
    public static LogSeverity Parse(string? text, LogSeverity fallback = LogSeverity.Info) =>
        TryParse(text, out var severity) ? severity : fallback;

    /// <summary>Numeric rank, for threshold comparisons.</summary>
    public static int Order(string? text) => (int)Parse(text);
}
